File: src/main/java/io/zeitwert/ddd/persistence/jooq/base/AggregatePersistenceProviderBase.java
package io.zeitwert.ddd.persistence.jooq.base;

import java.util.HashMap;
import java.util.Map;

import org.jooq.DSLContext;

import io.zeitwert.ddd.aggregate.model.Aggregate;
import io.zeitwert.ddd.aggregate.model.AggregateRepository;
import io.zeitwert.ddd.app.service.api.AppContext;

public abstract class AggregatePersistenceProviderBase<A extends Aggregate> {

	private final Map<String, Object> dbConfigMap = new HashMap<>();
	private final DSLContext dslContext;
	private final Class<? extends AggregateRepository<A, ?>> repoIntfClass;

	public AggregatePersistenceProviderBase(
			Class<? extends AggregateRepository<A, ?>> repoIntfClass,
			Class<? extends Aggregate> baseClass,
			DSLContext dslContext) {
		this.repoIntfClass = repoIntfClass;
		this.dslContext = dslContext;
	}

	public final Map<String, Object> dbConfigMap() {
		return this.dbConfigMap;
	}

	public final DSLContext dslContext() {
		return this.dslContext;
	}

	public final AggregateRepository<A, ?> getRepository() {
		return AppContext.getInstance().getBean(this.repoIntfClass);
	}

}

File: fm-dddrive/src/main/java/io/dddrive/ddd/model/base/AggregatePersistenceProviderBase.kt
package io.dddrive.ddd.model.base

import io.dddrive.ddd.model.Aggregate
import io.dddrive.ddd.model.AggregatePersistenceProvider
import io.dddrive.ddd.model.RepositoryDirectory
import io.dddrive.ddd.model.RepositoryDirectorySPI

abstract class AggregatePersistenceProviderBase<A : Aggregate>(
	intfClass: Class<A>,
) : AggregatePersistenceProvider<A> {

	val directory: RepositoryDirectory get() = RepositoryDirectory.instance

	init {
		(directory as RepositoryDirectorySPI).addPersistenceProvider(intfClass, this)
	}

}
